[contain-intrinsic-size] Support "auto none"

rwlbuis opened the pull request:
#### 242f3ae3216ce12c84cdf938b072227b5da42cc3
<pre>
[contain-intrinsic-size] Support &quot;auto none&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=258618">https://bugs.webkit.org/show_bug.cgi?id=258618</a>

Reviewed by Tim Nguyen.

See:
<a href="https://github.com/w3c/csswg-drafts/pull/8989">https://github.com/w3c/csswg-drafts/pull/8989</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-014-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-014.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/parsing/contain-intrinsic-size-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/parsing/contain-intrinsic-size-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/parsing/contain-intrinsic-size-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/parsing/contain-intrinsic-size-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/parsing/contain-intrinsic-size-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/parsing/contain-intrinsic-size-valid.html:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForContainIntrinsicSize):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeContainIntrinsicSize):
* Source/WebCore/dom/Document.cpp:
(WebCore::CallbackForContainIntrinsicSize):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::explicitIntrinsicInnerWidth const):
(WebCore::RenderBox::explicitIntrinsicInnerHeight const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::containIntrinsicHeightHasAuto const):
(WebCore::RenderStyle::containIntrinsicWidthHasAuto const):
(WebCore::RenderStyle::hasAutoLengthContainIntrinsicSize const):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueContainIntrinsicWidth):
(WebCore::Style::BuilderCustom::applyValueContainIntrinsicHeight):

Canonical link: <a href="https://commits.webkit.org/265617@main">https://commits.webkit.org/265617@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de3bf5e991b3d2d19586148d079a484cc05ac99a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11415 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11619 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11980 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13060 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10865 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11436 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14004 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11601 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13778 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11578 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12449 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9672 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13480 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9741 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10361 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17526 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10812 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10515 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13710 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10915 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8988 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10090 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10241 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2737 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14363 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10771 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -8097,16 +8097,17 @@ RefPtr<CSSValue> consumeContain(CSSParserTokenRange& range)
+        if (autoValue)
+            return CSSValueList::createSpaceSeparated(autoValue.releaseNonNull(), noneValue.releaseNonNull());
+        return noneValue;
     }
```
I'd rewrite this parsing function to something like this. Also we can use `CSSValuePair` instead of `CSSValueList` here.

```cpp
RefPtr<CSSPrimitiveValue> autoValue;
if (autoValue = consumeIdent<CSSValueAuto>(range)) {
    if (range.atEnd())
        return nullptr;
}

if (auto noneValue = consumeIdent<CSSValueNone>(range)) {
    if (autoValue)
        return CSSValuePair::create(autoValue.releaseNonNull(), noneValue.releaseNonNull());
    return noneValue;
}

if (auto lengthValue = consumeLength(range, HTMLStandardMode, ValueRange::NonNegative)) {
    if (autoValue)
        return CSSValuePair::create(autoValue.releaseNonNull(), lengthValue.releaseNonNull());
    return lengthValue;
}
```

rwlbuis replied:
Done.

nt1m commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -2101,10 +2101,14 @@ inline void BuilderCustom::applyValueContainIntrinsicWidth(BuilderState& builder
-    style.setContainIntrinsicWidthType(ContainIntrinsicSizeType::AutoAndLength);
-    auto lengthValue = downcast<CSSPrimitiveValue>(list.item(1))->computeLength<Length>(builderState.cssToLengthConversionData().copyWithAdjustedZoom(1.0f));
-    style.setContainIntrinsicWidth(lengthValue);
+    ASSERT(downcast<CSSPrimitiveValue>(list.item(1))->isLength() || downcast<CSSPrimitiveValue>(list.item(1))->valueID() == CSSValueNone);
```
I think this specific ASSERT is redundant as you do the same check in the if. You could move the length assert into else branch.

rwlbuis replied:
Removed.

nt1m commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -2101,10 +2101,14 @@ inline void BuilderCustom::applyValueContainIntrinsicWidth(BuilderState& builder
     auto& list = downcast<CSSValueList>(value);
     ASSERT(list.length() == 2);
```
This can go away if you use CSSValuePair.

rwlbuis replied:
Nice, fixed.

sammygill commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -416,9 +416,9 @@ static void CallbackForContainIntrinsicSize(const Vector<Ref<ResizeObserverEntry
 
             auto contentBoxSize = entry->contentBoxSize().at(0);
-            if (box->style().containIntrinsicLogicalWidthType() == ContainIntrinsicSizeType::AutoAndLength)
+            if (box->style().containIntrinsicWidthHasAuto())
```
It appears to me that we are going from checking the logical properties to the physical ones. Is this ok or should we stick to checking the logical properties and make this `containIntrinsicLogicalWidthHasAuto`?

nt1m replied:
This is a good catch, @rwlbuis can you take a look? we were taking in account writing mode before here, now we're not.

rwlbuis replied:
Good point indeed, I created https://github.com/WebKit/WebKit/pull/15812 to address this.

sammygill commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -416,9 +416,9 @@ static void CallbackForContainIntrinsicSize(const Vector<Ref<ResizeObserverEntry
+            if (box->style().containIntrinsicWidthHasAuto())
                 target->setLastRememberedLogicalWidth(LayoutUnit(contentBoxSize->inlineSize()));
-            if (box->style().containIntrinsicLogicalHeightType() == ContainIntrinsicSizeType::AutoAndLength)
+            if (box->style().containIntrinsicHeightHasAuto())
```
Same thing here. Should this be `containIntrinsicLogicalHeightHasAuto`?

sammygill commented on LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/parsing/contain-intrinsic-size-valid.html:
```diff
@@ -42,6 +42,8 @@
 test_valid_value("contain-intrinsic-block-size", "1px");
 test_valid_value("contain-intrinsic-block-size", "2em");
 test_valid_value("contain-intrinsic-block-size", "auto 1px");
+
```
Extra line here?

rwlbuis committed:
[contain-intrinsic-size] Support "auto none"
https://bugs.webkit.org/show_bug.cgi?id=258618

Reviewed by Tim Nguyen.

See:
https://github.com/w3c/csswg-drafts/pull/8989

* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-014-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-014.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/parsing/contain-intrinsic-size-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/parsing/contain-intrinsic-size-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/parsing/contain-intrinsic-size-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/parsing/contain-intrinsic-size-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/parsing/contain-intrinsic-size-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/parsing/contain-intrinsic-size-valid.html:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForContainIntrinsicSize):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeContainIntrinsicSize):
* Source/WebCore/dom/Document.cpp:
(WebCore::CallbackForContainIntrinsicSize):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::explicitIntrinsicInnerWidth const):
(WebCore::RenderBox::explicitIntrinsicInnerHeight const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::containIntrinsicHeightHasAuto const):
(WebCore::RenderStyle::containIntrinsicWidthHasAuto const):
(WebCore::RenderStyle::hasAutoLengthContainIntrinsicSize const):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueContainIntrinsicWidth):
(WebCore::Style::BuilderCustom::applyValueContainIntrinsicHeight):

Canonical link: https://commits.webkit.org/265617@main
